cpp/dolfinx/fem/utils.cpp

```diff
@@ -277,3 +277,110 @@ fem::FunctionSpace fem::create_functionspace(
           mesh->comm(), layout, mesh->topology(), reorder_fn, *element)));
 }
 //-----------------------------------------------------------------------------
+std::map<int, std::vector<std::int32_t>>
+fem::compute_integration_domains(const fem::IntegralType integral_type,
+                                 const mesh::MeshTags<int>& meshtags)
+{
+  std::map<int, std::vector<std::int32_t>> integrals;
+
+  std::shared_ptr<const mesh::Mesh> mesh = meshtags.mesh();
+  const mesh::Topology& topology = mesh->topology();
+  const int tdim = topology.dim();
+  int dim = integral_type == IntegralType::cell ? tdim : tdim - 1;
+  if (dim != meshtags.dim())
+  {
+    throw std::runtime_error("Invalid MeshTags dimension: "
+                             + std::to_string(meshtags.dim()));
+  }
+
+  // Get mesh tag data
+  const std::vector<int>& tags = meshtags.values();
+  const std::vector<std::int32_t>& tagged_entities = meshtags.indices();
+  assert(topology.index_map(dim));
+  const auto entity_end
+      = std::lower_bound(tagged_entities.begin(), tagged_entities.end(),
+                         topology.index_map(dim)->size_local());
+
+  switch (integral_type)
+  {
+  case fem::IntegralType::cell:
+  {
+    for (auto c = tagged_entities.cbegin(); c != entity_end; ++c)
+    {
+      const std::size_t index = std::distance(tagged_entities.cbegin(), c);
+      integrals[tags[index]].push_back(*c);
+    }
+  }
+  break;
+  default:
+    mesh->topology_mutable().create_connectivity(dim, tdim);
+    mesh->topology_mutable().create_connectivity(tdim, dim);
+    auto f_to_c = topology.connectivity(tdim - 1, tdim);
+    assert(f_to_c);
+    auto c_to_f = topology.connectivity(tdim, tdim - 1);
+    assert(c_to_f);
+    switch (integral_type)
+    {
+    case IntegralType::exterior_facet:
+    {
+      // When a mesh is not ghosted by cell, it is not straightforward
+      // to distinguish between (i) exterior facets and (ii) interior
+      // facets that are on a partition boundary. If there are no
+      // ghost cells, build a set of owned facts that are ghosted on
+      // another process to help determine if a facet is on an
+      // exterior boundary.
+      assert(topology.index_map(tdim - 1));
+      const std::vector<std::int32_t> fwd_shared_facets
```

Comment:
I think we now have a function that determined exterior facets without relying on the ghosting type.